Math/NumberTheory/Moduli/Sqrt.hs

```diff
@@ -30,13 +27,63 @@ import GHC.Integer.GMP.Internals
 
 import Math.NumberTheory.Moduli.Chinese
 import Math.NumberTheory.Moduli.Jacobi
+import Math.NumberTheory.Powers.General (highestPower)
 import Math.NumberTheory.Primes.Sieve (sieveFrom)
+import Math.NumberTheory.Primes.Testing.Probabilistic (isPrime)
 import Math.NumberTheory.Utils (shiftToOddCount, splitOff)
 
+-- | Extended flag structure to check if square root is computed in finite field.
+data FieldCharacterictic = Invalid
+                         | Prime
+                         | PrimePower Integer Int
+  deriving Show
```

Comment:
This is a nice approach, but I'd rather separate validation of modulo from actual sqrt computations for the sake of performance. It is quite typical in applications to compute a multitude of square roots modulo same huge prime, and it is better to avoid rerunning `isPrime` on each call. 

One can introduce something like `data FieldCharacteristic = FieldCharacteristic (Prime Integer) Word`, keeping prime number and its power (possibly 1) and change `sqrt*` functions to accept `FieldCharacteristic` directly. 
E. g., `sqrtModList :: Integer -> FieldCharacteristic -> [Integer]`, `sqrtModMaybe :: Integer -> FieldCharacteristic -> Maybe Integer` and so on. 

What do you think about it?

Reply:
I agree.

So i see it as
```
data FieldCharacteristic = FieldCharacteristic (Prime Integer) Word

toFieldCharacteristic :: Integer -> FieldCharacteristic
``` in `Math.NumberTheory.Moduli.Class`
And corresponding body of builder, and signatures of `sqrt` functions.